File: Managed/NextTurn.UE.Runtime/CoreUObject/Package.cs
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

using System;
using NextTurn.UE.Annotations;

namespace Unreal
{
    public sealed class Package : Object
    {
        public static Package Any { get; } = default!;

        internal Package(IntPtr pointer) : base(pointer) { }

        public long FileLength => NativeMethods.GetFileLength(this.pointer);

        public Name FileName => NativeMethods.GetFileName(this.pointer);

        public Guid Guid => NativeMethods.GetGuid(this.pointer);

        private static new class NativeMethods
        {
            [ReadOffset]
            public static extern long GetFileLength(IntPtr package);

            [ReadOffset]
            public static extern Name GetFileName(IntPtr package);

            [ReadOffset]
            public static extern Guid GetGuid(IntPtr package);
        }
    }
}

File: Source/NextTurnRuntime/Private/CoreUObject/Package.cpp
// Copyright (c) NextTurn. All rights reserved.
// Licensed under the Apache License, Version 2.0.
// See LICENSE.txt in the project root for more information.

#include "NextTurnRuntime/Private/Exports.hpp"

#include "NextTurnRuntime/Private/Accessor.hpp"
#include "NextTurnRuntime/Public/Compiler.hpp"

#include "Runtime/Core/Public/CoreMinimal.h"

#include "Runtime/CoreUObject/Public/UObject/Package.h"

ENABLE_WARNINGS;

DEFINE_OFFSET_ACCESSOR(UPackage, Guid);

#define TYPE_EXPORT_MANAGED_TYPE_NAME Package

TYPE_EXPORT_START

const FMemberSymbol Members[] =
{
	EXPORT_OFFSET_PUBLIC(UPackage, FileLength, FileSize),
	EXPORT_OFFSET_PUBLIC(UPackage, FileName, FileName),
	EXPORT_OFFSET(UPackage, Guid, Guid),
};

TYPE_EXPORT_END

#undef TYPE_EXPORT_MANAGED_TYPE_NAME

RESTORE_WARNINGS;
